# Activity 3.1 - Recurrent Neural Networks

#### Objective(s):

This activity aims to introduce how to build a recurrent neural network

#### Intended Learning Outcomes (ILOs):
* Demonstrate how to build and train neural recurrent neural network
* Evaluate the score and accuracy of the recurrent neural network

#### Resources:
* Jupyter Notebook
* IMDB

#### Procedures
Load the necessary libraries

In [1]:
from __future__ import print_function
import keras
from keras.preprocessing import sequence
from keras.models import Sequential
from keras.layers import Dense, Embedding
from keras.layers import SimpleRNN
from keras.datasets import imdb
from keras import initializers

* Set the max_features to 20000
* Set the maximum length of a sequence
* Use the batch size of 32

In [2]:
max_features = 20000
maxlen = 30
batch_size = 32

Load the data

In [3]:

(x_train, y_train), (x_test, y_test) = imdb.load_data(num_words=max_features)
print(len(x_train), 'train sequences')
print(len(x_test), 'test sequences')

17464789/17464789 [==============================] - 0s 0us/step
25000 train sequences
25000 test sequences


 Truncates the sequences so that they are of the maximum length

In [4]:

x_train = sequence.pad_sequences(x_train, maxlen=maxlen)
x_test = sequence.pad_sequences(x_test, maxlen=maxlen)
print('x_train shape:', x_train.shape)
print('x_test shape:', x_test.shape)

x_train shape: (25000, 30)
x_test shape: (25000, 30)


In [6]:
#Check the example of sequence

In [7]:
x_train[123,:]

array([  219,   141,    35,   221,   956,    54,    13,    16,    11,
        2714,    61,   322,   423,    12,    38,    76,    59,  1803,
          72,     8, 10508,    23,     5,   967,    12,    38,    85,
          62,   358,    99], dtype=int32)

Build a recurrent neural network

In [8]:
rnn_hidden_dim = 5
word_embedding_dim = 50
model_rnn = Sequential()
model_rnn.add(Embedding(max_features, word_embedding_dim))  #This layer takes each integer in the sequence and embeds it in a 50-dimensional vector
model_rnn.add(SimpleRNN(rnn_hidden_dim,
                    kernel_initializer=initializers.RandomNormal(stddev=0.001),
                    recurrent_initializer=initializers.Identity(gain=1.0),
                    activation='relu',
                    input_shape=x_train.shape[1:]))

model_rnn.add(Dense(1, activation='sigmoid'))

In [9]:
rmsprop = keras.optimizers.RMSprop(lr = .0001)

model_rnn.compile(loss='binary_crossentropy',
              optimizer=rmsprop,
              metrics=['accuracy'])

In [11]:
model_rnn.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=20,
          validation_data=(x_test, y_test))

Epoch 1/20
782/782 [==============================] - 13s 17ms/step - loss: 0.2400 - accuracy: 0.9054 - val_loss: 0.5494 - val_accuracy: 0.7774
Epoch 2/20
782/782 [==============================] - 12s 15ms/step - loss: 0.2271 - accuracy: 0.9116 - val_loss: 0.5276 - val_accuracy: 0.7724
Epoch 3/20
782/782 [==============================] - 12s 15ms/step - loss: 0.2158 - accuracy: 0.9175 - val_loss: 0.5478 - val_accuracy: 0.7768
Epoch 4/20
782/782 [==============================] - 12s 15ms/step - loss: 0.2062 - accuracy: 0.9210 - val_loss: 0.5591 - val_accuracy: 0.7696
Epoch 5/20
782/782 [==============================] - 11s 14ms/step - loss: 0.1975 - accuracy: 0.9253 - val_loss: 0.5551 - val_accuracy: 0.7731
Epoch 6/20
782/782 [==============================] - 11s 14ms/step - loss: 0.1892 - accuracy: 0.9290 - val_loss: 0.6090 - val_accuracy: 0.7746
Epoch 7/20
782/782 [==============================] - 10s 13ms/step - loss: 0.1809 - accuracy: 0.9338 - val_loss: 0.6017 - val_accuracy:

Evaluate the model using the test set.

In [12]:
score, acc = model_rnn.evaluate(x_test, y_test,
                            batch_size=batch_size)
print('Test score:', score)
print('Test accuracy:', acc)

782/782 [==============================] - 2s 2ms/step - loss: 0.7036 - accuracy: 0.7482
Test score: 0.7036314606666565
Test accuracy: 0.7482399940490723


Interpret the result
- The results shows that the val_loss after 20 epochs is 70.36% and the val_accuracy is 74.82% with a accuracy of 95.73%.

#type your answer here

#### Supplementary Activity

- Prepare the data to use sequences of length 80 rather than length 30.  Did it improve the performance?
- Try different values of the "max_features".  Can you improve the performance?
- Try smaller and larger sizes of the RNN hidden dimension.  How does it affect the model performance?  How does it affect the run time?

In [14]:
max_features = 20000
maxlen = 80

(x_train, y_train), (x_test, y_test) = imdb.load_data(num_words=max_features)

x_train = sequence.pad_sequences(x_train, maxlen=maxlen)
x_test = sequence.pad_sequences(x_test, maxlen=maxlen)

In [15]:
rnn_hidden_dim = 5
word_embedding_dim = 50
model_rnn = Sequential()
model_rnn.add(Embedding(max_features, word_embedding_dim))
model_rnn.add(SimpleRNN(rnn_hidden_dim,
                    kernel_initializer=initializers.RandomNormal(stddev=0.001),
                    recurrent_initializer=initializers.Identity(gain=1.0),
                    activation='relu',
                    input_shape=x_train.shape[1:]))

model_rnn.add(Dense(1, activation='sigmoid'))

In [16]:
rmsprop = keras.optimizers.RMSprop(lr = .0001)

model_rnn.compile(loss='binary_crossentropy',
              optimizer=rmsprop,
              metrics=['accuracy'])

In [17]:

model_rnn.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=10,
          validation_data=(x_test, y_test))

Epoch 1/10
782/782 [==============================] - 22s 26ms/step - loss: 0.4648 - accuracy: 0.7631 - val_loss: 0.3607 - val_accuracy: 0.8382
Epoch 2/10
782/782 [==============================] - 21s 27ms/step - loss: 0.3121 - accuracy: 0.8656 - val_loss: 0.3760 - val_accuracy: 0.8311
Epoch 3/10
782/782 [==============================] - 19s 24ms/step - loss: 0.2727 - accuracy: 0.8866 - val_loss: 0.4304 - val_accuracy: 0.8118
Epoch 4/10
782/782 [==============================] - 22s 28ms/step - loss: 0.2454 - accuracy: 0.9012 - val_loss: 0.3601 - val_accuracy: 0.8445
Epoch 5/10
782/782 [==============================] - 19s 25ms/step - loss: 0.2221 - accuracy: 0.9130 - val_loss: 0.3901 - val_accuracy: 0.8306
Epoch 6/10
782/782 [==============================] - 20s 26ms/step - loss: 0.2037 - accuracy: 0.9214 - val_loss: 0.4212 - val_accuracy: 0.8373
Epoch 7/10
782/782 [==============================] - 21s 27ms/step - loss: 0.1870 - accuracy: 0.9302 - val_loss: 0.4078 - val_accuracy:

In [18]:
max_features = 5000
maxlen = 80

(x_train, y_train), (x_test, y_test) = imdb.load_data(num_words=max_features)

x_train = sequence.pad_sequences(x_train, maxlen=maxlen)
x_test = sequence.pad_sequences(x_test, maxlen=maxlen)

In [19]:
rnn_hidden_dim = 5
word_embedding_dim = 20
model_rnn = Sequential()
model_rnn.add(Embedding(max_features, word_embedding_dim))  #This layer takes each integer in the sequence
model_rnn.add(SimpleRNN(rnn_hidden_dim,
                    kernel_initializer=initializers.RandomNormal(stddev=0.001),
                    recurrent_initializer=initializers.Identity(gain=1.0),
                    activation='relu',
                    input_shape=x_train.shape[1:]))

model_rnn.add(Dense(1, activation='sigmoid'))

In [20]:
rmsprop = keras.optimizers.RMSprop(lr = .0001)

model_rnn.compile(loss='binary_crossentropy',
              optimizer=rmsprop,
              metrics=['accuracy'])

In [21]:
model_rnn.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=10,
          validation_data=(x_test, y_test))

Epoch 1/10
782/782 [==============================] - 18s 22ms/step - loss: 0.6932 - accuracy: 0.4980 - val_loss: 0.6932 - val_accuracy: 0.5000
Epoch 2/10
782/782 [==============================] - 16s 20ms/step - loss: 0.6932 - accuracy: 0.4970 - val_loss: 0.6931 - val_accuracy: 0.5000
Epoch 3/10
782/782 [==============================] - 16s 20ms/step - loss: 0.6932 - accuracy: 0.4981 - val_loss: 0.6932 - val_accuracy: 0.5000
Epoch 4/10
782/782 [==============================] - 16s 20ms/step - loss: 0.6932 - accuracy: 0.4978 - val_loss: 0.6931 - val_accuracy: 0.5000
Epoch 5/10
782/782 [==============================] - 16s 20ms/step - loss: 0.6932 - accuracy: 0.4965 - val_loss: 0.6931 - val_accuracy: 0.5000
Epoch 6/10
782/782 [==============================] - 18s 23ms/step - loss: 0.6932 - accuracy: 0.5014 - val_loss: 0.6932 - val_accuracy: 0.5000
Epoch 7/10
782/782 [==============================] - 16s 20ms/step - loss: 0.6932 - accuracy: 0.4987 - val_loss: 0.6931 - val_accuracy:

In [22]:
#running for more epoch
model_rnn.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=10,
          validation_data=(x_test, y_test))

Epoch 1/10
782/782 [==============================] - 15s 20ms/step - loss: 0.6932 - accuracy: 0.4990 - val_loss: 0.6931 - val_accuracy: 0.5000
Epoch 2/10
782/782 [==============================] - 15s 19ms/step - loss: 0.6932 - accuracy: 0.4906 - val_loss: 0.6931 - val_accuracy: 0.5000
Epoch 3/10
782/782 [==============================] - 15s 20ms/step - loss: 0.6932 - accuracy: 0.4999 - val_loss: 0.6931 - val_accuracy: 0.5000
Epoch 4/10
782/782 [==============================] - 18s 23ms/step - loss: 0.6932 - accuracy: 0.4977 - val_loss: 0.6931 - val_accuracy: 0.5000
Epoch 5/10
782/782 [==============================] - 16s 20ms/step - loss: 0.6932 - accuracy: 0.4996 - val_loss: 0.6932 - val_accuracy: 0.5000
Epoch 6/10
782/782 [==============================] - 16s 20ms/step - loss: 0.6932 - accuracy: 0.4990 - val_loss: 0.6931 - val_accuracy: 0.5000
Epoch 7/10
782/782 [==============================] - 16s 20ms/step - loss: 0.6932 - accuracy: 0.4983 - val_loss: 0.6931 - val_accuracy:

#### Conclusion

To conclude, recurrent neural networks (RNNs) represent an essential architecture for sequential data processing tasks. Since they can remember past events by using recurrent connections, they are good at activities like natural language processing, time series analysis, or speech recognition. Traditional RNNs have performed well, but suffered from the gradient disappearance issue, which limits their capacity to capture long-distance dependencies. Consequently, some new versions like the long short-term memory (LSTM) and gated recurrent units (GRUs) have been created to counter this by providing better memory retention and training stability. While more studies are still continuing in the area of refining RNN networks and training approaches, they still serve as a base for conduct deep learning in the future.